File: 25-10-2025_Minimum_Steps_to_Halve_Sum.cpp
class Solution {
  public:
    int minOperations(vector<int>& arr) {
        // code here
        double sum = 0;
        priority_queue<double> pq;
        
        for (int x : arr) {
            sum += x;
            pq.push(x);
        }
        
        double target = sum / 2.0;
        double curr = sum;
        int count = 0;
        
        while (curr > target) {
            double x = pq.top(); pq.pop();
            double half = x / 2.0;
            curr -= half;
            pq.push(half);
            count++;
        }
        
        return count;
    }

};

File: 25-10-2025_Minimum_Steps_to_Halve_Sum.py
class Solution:
    def minOperations(self, arr):
        total = sum(arr)
        target = total / 2
        curr = total
        count = 0

        # Python has a min-heap by default, so we store negatives for max-heap behavior
        max_heap = [-x for x in arr]
        heapq.heapify(max_heap)

        while curr > target:
            largest = -heapq.heappop(max_heap)  # get the largest element
            half = largest / 2
            curr -= half
            heapq.heappush(max_heap, -half)     # push halved value back
            count += 1

        return count
